File: Graph/go/has_cycle.go
package graph

// HasCycle detects a cycle in a directed graph.
func HasCycle(g Graph) bool {
	visiting := map[string]struct{}{}
	visited := map[string]struct{}{}

	for node := range g {
		if cycleDFS(g, node, visiting, visited) {
			return true
		}
	}
	return false
}

func cycleDFS(g Graph, node string, visiting, visited map[string]struct{}) bool {
	if _, done := visited[node]; done {
		return false
	}
	if _, active := visiting[node]; active {
		return true
	}

	visiting[node] = struct{}{}
	for _, neighbor := range g[node] {
		if cycleDFS(g, neighbor, visiting, visited) {
			return true
		}
	}
	delete(visiting, node)
	visited[node] = struct{}{}
	return false
}

File: Graph/php/hasCycle.php
<?php

declare(strict_types=1);

function hasCycle(array $graph): bool
{
    $visiting = [];
    $visited = [];

    foreach ($graph as $node => $_) {
        if (detectCycle($graph, (string) $node, $visiting, $visited)) {
            return true;
        }
    }

    return false;
}

function detectCycle(array $graph, string $node, array &$visiting, array &$visited): bool
{
    if (isset($visited[$node])) {
        return false;
    }
    if (isset($visiting[$node])) {
        return true;
    }

    $visiting[$node] = true;
    foreach ($graph[$node] ?? [] as $neighbor) {
        if (detectCycle($graph, (string) $neighbor, $visiting, $visited)) {
            return true;
        }
    }

    unset($visiting[$node]);
    $visited[$node] = true;
    return false;
}
